### DISCUSSION TITLE
[Best Practices] ConstraintViolationList & i18n

### DISCUSSION THREAD

**Discussion by bpolaszek** [api-platform/core#5366] [Q&A]
Hello there,

Consider the following entity:
```php
use ApiPlatform\Core\Annotation\ApiResource;
use Symfony\Component\Validator\Constraints as Assert;

#[ApiResource]
class SomeResource {
   #[Assert\Count(max: 10)]
   public array $items;
}
```

Whenever an API consumer (the JS app) populates `items` with more than 10 elements, it gets the following response:
```json
{
  "@context": "/api/contexts/ConstraintViolationList",
  "@type": "ConstraintViolationList",
  "hydra:title": "An error occurred",
  "hydra:description": "items: This collection should contain 10 elements or less.",
  "violations": [
    {
      "propertyPath": "items",
      "message": "This collection should contain 10 elements or less.",
      "code": "756b1212-697c-468d-a9ad-50dd783bb169"
    }
  ]
}
```

However, for non-english speakers, some validation error messages are pretty hard to understand (and make the reading inconsistent as different languages are mixed). 
So, I thought about mapping `violations[*].code` to a set of xlated messages on the front-end side, except I cannot easily access the variable(s) that triggered the violation (here: `max` with a value of `10`).

What's the best way to tackle this? I assume I'm not the 1st one to face that issue.

Thanks,
Ben

**Comment by alanpoulain** [comment]
Hello,
Symfony validation messages are all translated in a lot of languages.
If the locale is correctly set, the message will be translated.
If you really want to access the parameters, we could add it optionally in the `AbstractConstraintViolationListNormalizer` but I'm not sure it's really needed.
You can also decorate the `ConstraintViolationListNormalizer` to add the logic yourself.

**Reply by bpolaszek** [reply]
Hey @alanpoulain,

Thanks for the quick reply.

Looks like it doesn't care about the locale, as `ConstraintViolationList` responses are always in English (not only on my side, but I also have customers complaints about this).

Just checked against Chrome devtools, which properly sends `Accept-Language: fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7`. 
Is there something I misconfigured on the API side?

**Reply by alanpoulain** [reply]
You probably need to configure Symfony as well: https://symfony.com/blog/new-in-symfony-5-4-language-negotiation.

**Reply by bpolaszek** [reply]
Oh, didn't know about this one.

Unfortunately, it still doesn't work either. Could it be related to the API-Platform version I'm using (2.6.8 - forgot to mention that earlier)?

**Reply by alanpoulain** [reply]
It should probably work if you are using the right Symfony version.

**Reply by bpolaszek** [reply]
I actually do (5.4.17). 🤷 

Will dig further and post updates here for the records, thanks for your precious help. 🙏

**Comment by norkunas** [comment]
> except I cannot easily access the variable(s) that triggered the violation (here: max with a value of 10).

As a workaround to avoid customizing serialization you could set your `max` also in `payload` option and access it, just configure `api_platform.validator. serialize_payload_fields: ~` option to expose it

**Reply by bpolaszek** [reply]
That's what I tried at first but it didn't work.

```yaml
api_platform:
    validator:
        serialize_payload_fields:
            - max
            - limit
```

Same response produced as above.

**Reply by norkunas** [reply]
Did you define `max` on `Assert\Count` payload? Meaning `Assert\Count(max: 10, payload: ['max' => 10])`

**Reply by bpolaszek** [reply]
Oh, no I didn't - I thought values would be bound directly. 

It works now! I'll probably use that if I don't manage to use @alanpoulain's solution (which would be way easier, actually...).

Thanks!
Ben
